### ISSUE TITLE
Check if the browser extension is installed 

### ISSUE THREAD

**Issue by shotgunofdeath** [airgap-it/beacon-sdk#12]
Hello there!

Could someone, please, tell if there is a way to check if the Beacon extension is installed before using any functionality of the SDK? For example, I would really like to check if a user has the extension installed in their browser before sending the requestPermissions() request.

Thank you in advance!

**Comment by AndreasGassmann** [comment]
First of all, could you please tell me the use case for this? Our SDK has an automatic fallback to the P2P approach to directly connect to the users mobile wallet. So the `requestPermissions()` can always be called, even if the user doesn't have the extension installed. In fact, there are many people who dislike to install browser extension (for good reasons), so they would expect the P2P pairing to be initiated on a beacon-compatible website.

That being said, there is no way to directly check if a chrome extension is installed. What you can do (and this is how we do it internally) is sending a "ping" message to all extensions and listen for a "pong". This means the check is asynchronous. Internally, we assume that if we did not receive an answer after 200ms, then the extension is not installed. The relevant code can be found here: https://github.com/airgap-it/beacon-sdk/blob/master/src/clients/client/Client.ts#L117

```
// You can use the following method to check for the "ping/pong" message.
// This promise will be pending until a "pong" is received. So you have define a time, eg. 200ms, when you decide to cancel and assume the extension is not installed.
PostMessageTransport.isAvailable().then(isAvailable => {})
```

In `beacon-sdk v2` we will provide a helper method that makes this easier. I will update this ticket once v2 is released.

**Comment by shotgunofdeath** [comment]
Thank you for the reply!

I'll try to explain our particular situation. I might be wrong regarding some specific behaviour of the SDK as I'm new to it, if so, correct me, please)

So, we're planning to build a dApp with JWT-authorisation and getting acquainted with the SDK. The idea is to use a public key of an active account (which we can get from Beacon) to validate a user's auth request on our back-end. 

At first, I thought that I could get that public key from a response of the requestPermissions() request, but it only returned a response body if the Beacon extension had been installed and returned nothing otherwise, moreover, there was no public key in the returned response body.

Then, I figured out that the SDK used the local storage of a browser for keeping some info, specifically, I was able to find there an array of accounts and an active account identifier. This is what we need, BUT, as far as I understand, accounts info is stored in the local storage of the browser only if the Beacon extension is installed. Because of that I think that we have to check if a user has the extension before any iteraction with the SDK, so that we can warn them, that they need the extension to continue using our app.

P.S. I've already found your implementation in the source code. I'll try use that for our puposes.

**Comment by AndreasGassmann** [comment]
Thanks for the explanations. So first, let me start with some basic beacon concepts.

The `beacon-sdk` automatically handles connections to both browser extensions and wallets directly (through a p2p network). You can try this out by going to our example page https://walletbeacon.io. If you have the extension installed and you start any action, the extension will be triggered. But if you don't have it installed (you can open it in inkognito mode for example), then you will see a QR popping up. This QR can be scanned in a beacon-compatible wallet (at the moment only AirGap Wallet supports it, but others like Galleon and Magma will integrate it as well). If the user scans the QR, a secure connection is established through a peer-to-peer network. So if you then invoke an action on the DApp side, the request gets sent directly to the wallet - no extension needed.

So there are basically 3 cases:
1. User has extension installed => You know that he can handle beacon requests
2. User doesn't have an extension => The user might be using a beacon-compatible wallet and is waiting for the QR to scan
3. User doesn't have an extension => The user might not know what this is and dismisses the QR

In cases 1 and 2, the `requestPermissions()` promise will eventually resolve with the permissions the users give you (in case 2. this might take some time, because he needs to scan the QR with his wallet, etc.).

The third case is trickier. We've actually not thought about handling this and you can currently not detect it because the promise will just never resolve if the user doesn't scan the QR.

What I'm thinking now is that we will add a "cancel" callback if the user decides to dismiss the QR popup, which will then reject the original promise. That would indicate to you that someone can't or doesn't want to handle beacon requests. Would that solve your issue?

So to summarise: Simply because someone doesn't have an extension, it doesn't mean he can't handle beacon requests. And you shouldn't "exclude" users who want to do the QR-scanning approach, because there are people (like me) who prefer not having browser extensions installed.

Now to your second point regarding the JWT authentication. I don't know exactly what you are trying to do, but I just wanted to let you know that "knowing" the public key is not enough to authenticate someone. The public key (of revealed accounts) for every address is available on the blockchain. If you want to authenticate someone as the owner of an address, you will have to use the `signPayload()` method to have the wallet sign a challenge that your server provides. Then your server can use challenge+signature+publicKey to verify the owner.

Regarding the publicKey not being in the response of `requestPermissions()`, I'll create create an issue for it and add it to the response.

**Comment by AndreasGassmann** [comment]
I added 2 new issues (https://github.com/airgap-it/beacon-sdk/issues/13, https://github.com/airgap-it/beacon-sdk/issues/14) for my suggested changes, please let me know what you think. I'll leave this issue open as a reminder that this behaviour should be documented because others might have the same question.

**Comment by shotgunofdeath** [comment]
Really appreciate your help! I think we've solved our problem.

I understand the general concept of the Beacon SDK, specifically that it should work both ways: with or without the Chrome extension and I think it's gerat. 

I've never thought about case 3 (**User doesn't have an extension => The user might not know what this is and dismisses the QR**) as well, but now I am)))

Getting a public key in the response will be very helpful! No need to address the local storage.

Also, we're aware that a public key only is not enough for authenticating and we are planning to implement a scheme close to what you've mentioned above. It's just the first thing we need at the moment for testing and, alas, we couldn't get it in the case 2 (**User doesn't have an extension => The user might be using a beacon-compatible wallet and is waiting for the QR to scan**). That's why I thought about checking if the extension was installed first. Fortunately, now it's clear what the problem was. No initial check is needed as you've said in your first comment (but still can be a useful feature to my mind). I'll leave the use case here, maybe it'll be interesting to someone. 

So, I've somehow managed to launch and use AirGap Wallet wthout importing any Tezos addresses from the AirGap Vault (Android). I had some problems with the "Install AirGap Vault" button in the wallet app - it just didn't work. So, I installed AirGap Vault manually, created an address but didn't sync it with the wallet, nevertheless, I could still use the QR-scanner of the wallet. 

As the result, when I tried connect to our test dApp with the Beacon QR-code window, the connection window of the wallet appeared on the phone screen with no checkboxes for permissions, but with the buttons "Done" and "Connect". Pressing "Connect" didn't do anything (or did not enough, not sure here) and the requestPermissions() promise was never resolved (the **/sync?timeout** request was repeated indefinetely) and we couldn't get any permissions. Don't know if it's an intended behaviour, anyway, I think it's worth sharing.

To sum up, the problem wasn't connected with the SDK itself, but rather with the AirGap Wallet and its usage with no addresses. But maybe the SDK should also somehow react to this case.

Thank you again! Your comments have helped me to go in the right direction to the solution.

**Comment by AndreasGassmann** [comment]
We have to improve the error handling (and fix the button that is apparently broken), but the reason the permission request is displayed like that and not working is because you don't have an address in AirGap Wallet. If you sync an account over from AirGap Vault, then it should work.

**Comment by shotgunofdeath** [comment]
You're right and now it works. but still these repeated requests (**https://matrix.papers.tech/_matrix/client/r0/sync?timeout...**) don't seem right. Or maybe I am doing something wrong. Regardless, I think that this problem is out of scope of this issue.

![pending-sync-request](https://user-images.githubusercontent.com/42133844/94005586-27f5b000-fda7-11ea-8cde-c71e8e4c7698.png)

**Comment by AndreasGassmann** [comment]
Those requests are expected. It is connecting to the beacon network and listening for new messages that might arrive for us (long-polling).

**Comment by shotgunofdeath** [comment]
OK, I see. Thank you very much for your time. Really appreciate the help!
